common/growing_range_test.cpp

```diff
@@ -0,0 +1,47 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "common/growing_range.h"
+
+#include <gtest/gtest.h>
+
+#include <vector>
+
+namespace Carbon {
+namespace {
+
+TEST(GrowingRangeTest, TestUnchanged) {
+  std::vector<int> v = {1, 2, 3, 4, 5};
+  int k = 0;
+  for (int n : GrowingRange(v)) {
+    EXPECT_EQ(n, ++k);
+  }
+}
+
+TEST(GrowingRangeTest, TestGrowWithRealloc) {
+  std::vector<int> expected = {3, 2, 2, 2, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0};
+
+  std::vector<int> v;
+  v.reserve(1);
+  v.push_back(3);
+  EXPECT_LT(v.capacity(), expected.size());
+
+  int i = 0;
+  for (int n : GrowingRange(v)) {
+    // Append n copies of n - 1.
+    v.insert(v.end(), n, n - 1);
+    EXPECT_EQ(n, expected[i++]);
+  }
+}
+
+TEST(GrowingRangeTest, TestNoReference) {
+  std::vector<int> v;
+  for (decltype(auto) elem : GrowingRange(v)) {
```

Comment:
Why the `decltype` here?

Reply:
The point is to detect whether the element would be a reference; `decltype(auto)` will deduce as a reference if the element is a reference, whereas `auto` is always a non-reference type.

Reply:
That's subtle for me personally; would it work to do something like:
```
static_assert(std::same_as<decltype(*GrowingRange(v).begin()), int>);
```

...or, maybe a comment?